overlord/snapshotstate/backend/backend.go

```diff
@@ -157,6 +157,60 @@ func Filename(snapshot *client.Snapshot) string {
 	return filepath.Join(dirs.SnapshotsDir, fmt.Sprintf("%d_%s_%s_%s.zip", snapshot.SetID, snapshot.Snap, snapshot.Version, snapshot.Revision))
 }
 
+// EstimateSnapshotSize calculates estimated size of the snapshot.
+// XXX: this doesn't take users into account yet, because it is used only
+// for automatic snapshots at the moment.
+func EstimateSnapshotSize(si *snap.Info) (int64, error) {
+	var total int64
+	calculateSize := func(path string, finfo os.FileInfo, err error) error {
+		if finfo.Mode().IsRegular() {
+			total += finfo.Size()
+		}
+		return err
+	}
+
+	// helper function for visiting revisionDir and common dir that is
+	// at the same level as revisionDir.
+	visitDataDir := func(revisionDir string) error {
+		parent := filepath.Dir(revisionDir)
+		exists, isDir, err := osutil.DirExists(parent)
+		if err != nil {
+			return err
+		}
+		if !(exists && isDir) {
+			return nil
+		}
+
+		exists, isDir, err = osutil.DirExists(revisionDir)
+		if err != nil {
+			return err
+		}
+		if exists && isDir {
+			filepath.Walk(revisionDir, calculateSize)
+		}
+
+		common := filepath.Join(parent, "common")
+		exists, isDir, err = osutil.DirExists(common)
+		if err != nil {
+			return err
+		}
+		if exists && isDir {
+			filepath.Walk(common, calculateSize)
+		}
+		return nil
+	}
+
+	if err := visitDataDir(si.DataDir()); err != nil {
+		return 0, err
+	}
```

Comment:
Maybe this could be simplified a little like so:
```
	visitDir := func(dir string) error {
		exists, isDir, err = osutil.DirExists(dir)
		if err != nil {
			return err
		}
		if !exists || !isDir {
			return nil
		}
		return filepath.Walk(dir, calculateSize)
	}

	for _, dir : = range []string{si.DataDir(), si.CommonDataDir()} {
		if err := visitDir(dir); err != nil {
			return err
		}
	}
```

Reply:
Done.